File: src/riscv_sim/cpu.py
"""The CPU execution engine: applies decoded instructions to machine state."""

import enum

from .bits import to_signed, to_unsigned
from .decoder import Instruction, decode
from .errors import IllegalInstructionError, MemoryAccessError, SimulatorError
from .memory import Memory
from .registers import RegisterFile


class Status(enum.Enum):
    RUNNING = "running"
    HALTED = "halted"
    ERROR = "error"


class CPU:
    """RV32I CPU state and execution engine."""

    def __init__(self, memory: Memory | None = None, pc: int = 0) -> None:
        self.memory = memory if memory is not None else Memory()
        self.regs = RegisterFile()
        self.pc = pc
        self.status = Status.RUNNING
        self.halt_reason: str | None = None
        self.exit_code = 0
        self.step_count = 0

    def step(self) -> Instruction:
        """Fetch, decode, and execute a single instruction.

        Returns the decoded instruction. If a fault occurs, ``status`` is set
        to ``Status.ERROR`` and ``halt_reason`` is populated; the caller
        should stop stepping once ``status`` is no longer ``RUNNING``.
        """
        try:
            raw = self.memory.read_word(self.pc)
            instr = decode(raw)
            self._execute(instr)
            self.step_count += 1
            return instr
        except SimulatorError as exc:
            self.status = Status.ERROR
            self.halt_reason = str(exc)
            raise

    def run(self, max_steps: int | None = 100_000, on_step=None) -> None:
        """Run until halted, faulted, or ``max_steps`` is reached."""
        steps = 0
        while self.status == Status.RUNNING:
            if max_steps is not None and steps >= max_steps:
                self.status = Status.ERROR
                self.halt_reason = f"exceeded max_steps ({max_steps}); possible infinite loop"
                break
            instr = self.step()
            if on_step is not None:
                on_step(instr)
            steps += 1

    def _execute(self, instr: Instruction) -> None:
        pc_next = self.pc + 4
        rs1 = self.regs.read(instr.rs1)
        rs2 = self.regs.read(instr.rs2)
        rs1_s = self.regs.read_signed(instr.rs1)
        rs2_s = self.regs.read_signed(instr.rs2)
        m = instr.mnemonic

        if m == "add":
            self.regs.write(instr.rd, rs1 + rs2)
        elif m == "sub":
            self.regs.write(instr.rd, rs1 - rs2)
        elif m == "sll":
            self.regs.write(instr.rd, rs1 << (rs2 & 0x1F))
        elif m == "slt":
            self.regs.write(instr.rd, int(rs1_s < rs2_s))
        elif m == "sltu":
            self.regs.write(instr.rd, int(rs1 < rs2))
        elif m == "xor":
            self.regs.write(instr.rd, rs1 ^ rs2)
        elif m == "srl":
            self.regs.write(instr.rd, rs1 >> (rs2 & 0x1F))
        elif m == "sra":
            self.regs.write(instr.rd, to_unsigned(rs1_s >> (rs2 & 0x1F)))
        elif m == "or":
            self.regs.write(instr.rd, rs1 | rs2)
        elif m == "and":
            self.regs.write(instr.rd, rs1 & rs2)

        elif m == "addi":
            self.regs.write(instr.rd, rs1 + instr.imm)
        elif m == "slti":
            self.regs.write(instr.rd, int(rs1_s < instr.imm))
        elif m == "sltiu":
            self.regs.write(instr.rd, int(rs1 < to_unsigned(instr.imm)))
        elif m == "xori":
            self.regs.write(instr.rd, rs1 ^ to_unsigned(instr.imm))
        elif m == "ori":
            self.regs.write(instr.rd, rs1 | to_unsigned(instr.imm))
        elif m == "andi":
            self.regs.write(instr.rd, rs1 & to_unsigned(instr.imm))
        elif m == "slli":
            self.regs.write(instr.rd, rs1 << instr.imm)
        elif m == "srli":
            self.regs.write(instr.rd, rs1 >> instr.imm)
        elif m == "srai":
            self.regs.write(instr.rd, to_unsigned(rs1_s >> instr.imm))

        elif m in ("lb", "lh", "lw", "lbu", "lhu"):
            address = to_unsigned(rs1 + instr.imm)
            if m == "lb":
                value = self.memory.read_byte(address, signed=True)
            elif m == "lh":
                value = self.memory.read_half(address, signed=True)
            elif m == "lw":
                value = self.memory.read_word(address)
            elif m == "lbu":
                value = self.memory.read_byte(address, signed=False)
            else:
                value = self.memory.read_half(address, signed=False)
            self.regs.write(instr.rd, value)

        elif m in ("sb", "sh", "sw"):
            address = to_unsigned(rs1 + instr.imm)
            if m == "sb":
                self.memory.write_byte(address, rs2)
            elif m == "sh":
                self.memory.write_half(address, rs2)
            else:
                self.memory.write_word(address, rs2)

        elif m in ("beq", "bne", "blt", "bge", "bltu", "bgeu"):
            taken = (
                (m == "beq" and rs1 == rs2)
                or (m == "bne" and rs1 != rs2)
                or (m == "blt" and rs1_s < rs2_s)
                or (m == "bge" and rs1_s >= rs2_s)
                or (m == "bltu" and rs1 < rs2)
                or (m == "bgeu" and rs1 >= rs2)
            )
            if taken:
                pc_next = to_unsigned(self.pc + instr.imm)

        elif m == "lui":
            self.regs.write(instr.rd, instr.imm)
        elif m == "auipc":
            self.regs.write(instr.rd, to_unsigned(self.pc + instr.imm))

        elif m == "jal":
            self.regs.write(instr.rd, pc_next)
            pc_next = to_unsigned(self.pc + instr.imm)
        elif m == "jalr":
            self.regs.write(instr.rd, pc_next)
            pc_next = to_unsigned((rs1 + instr.imm) & ~0x1)

        elif m == "fence":
            pass
        elif m == "ecall":
            self.status = Status.HALTED
            self.halt_reason = "ecall"
            self.exit_code = to_signed(self.regs.read(10))
        elif m == "ebreak":
            self.status = Status.HALTED
            self.halt_reason = "ebreak"
            self.exit_code = to_signed(self.regs.read(10))
        else:
            raise IllegalInstructionError(f"unimplemented mnemonic '{m}' (0x{instr.raw:08x})")

        self.pc = pc_next

File: src/cpu.cpp
#include "riscv_sim/cpu.hpp"

#include "riscv_sim/bits.hpp"
#include "riscv_sim/errors.hpp"

namespace riscv_sim {

Instruction CPU::step() {
    try {
        std::uint32_t raw = memory.read_word(pc);
        Instruction instr = decode(raw);
        execute(instr);
        ++step_count;
        return instr;
    } catch (const SimulatorError& exc) {
        status = Status::Error;
        halt_reason = exc.what();
        throw;
    }
}

void CPU::run(std::optional<long> max_steps, const std::function<void(const Instruction&)>& on_step) {
    long steps = 0;
    while (status == Status::Running) {
        if (max_steps.has_value() && steps >= *max_steps) {
            status = Status::Error;
            halt_reason = "exceeded max_steps (" + std::to_string(*max_steps) + "); possible infinite loop";
            break;
        }
        Instruction instr = step();
        if (on_step) {
            on_step(instr);
        }
        ++steps;
    }
}

void CPU::execute(const Instruction& instr) {
    std::uint32_t pc_next = pc + 4;
    std::uint32_t rs1 = regs.read(instr.rs1);
    std::uint32_t rs2 = regs.read(instr.rs2);
    std::int32_t rs1_s = regs.read_signed(instr.rs1);
    std::int32_t rs2_s = regs.read_signed(instr.rs2);
    const std::string& m = instr.mnemonic;

    if (m == "add") {
        regs.write(instr.rd, rs1 + rs2);
    } else if (m == "sub") {
        regs.write(instr.rd, rs1 - rs2);
    } else if (m == "sll") {
        regs.write(instr.rd, rs1 << (rs2 & 0x1F));
    } else if (m == "slt") {
        regs.write(instr.rd, rs1_s < rs2_s ? 1 : 0);
    } else if (m == "sltu") {
        regs.write(instr.rd, rs1 < rs2 ? 1 : 0);
    } else if (m == "xor") {
        regs.write(instr.rd, rs1 ^ rs2);
    } else if (m == "srl") {
        regs.write(instr.rd, rs1 >> (rs2 & 0x1F));
    } else if (m == "sra") {
        regs.write(instr.rd, to_unsigned(rs1_s >> (rs2 & 0x1F)));
    } else if (m == "or") {
        regs.write(instr.rd, rs1 | rs2);
    } else if (m == "and") {
        regs.write(instr.rd, rs1 & rs2);

    } else if (m == "addi") {
        regs.write(instr.rd, rs1 + instr.imm);
    } else if (m == "slti") {
        regs.write(instr.rd, rs1_s < instr.imm ? 1 : 0);
    } else if (m == "sltiu") {
        regs.write(instr.rd, rs1 < to_unsigned(instr.imm) ? 1 : 0);
    } else if (m == "xori") {
        regs.write(instr.rd, rs1 ^ to_unsigned(instr.imm));
    } else if (m == "ori") {
        regs.write(instr.rd, rs1 | to_unsigned(instr.imm));
    } else if (m == "andi") {
        regs.write(instr.rd, rs1 & to_unsigned(instr.imm));
    } else if (m == "slli") {
        regs.write(instr.rd, rs1 << instr.imm);
    } else if (m == "srli") {
        regs.write(instr.rd, rs1 >> instr.imm);
    } else if (m == "srai") {
        regs.write(instr.rd, to_unsigned(rs1_s >> instr.imm));

    } else if (m == "lb" || m == "lh" || m == "lw" || m == "lbu" || m == "lhu") {
        std::uint32_t address = to_unsigned(rs1 + instr.imm);
        std::int32_t value;
        if (m == "lb") {
            value = memory.read_byte(address, /*signed_=*/true);
        } else if (m == "lh") {
            value = memory.read_half(address, /*signed_=*/true);
        } else if (m == "lw") {
            value = static_cast<std::int32_t>(memory.read_word(address));
        } else if (m == "lbu") {
            value = memory.read_byte(address, /*signed_=*/false);
        } else {
            value = memory.read_half(address, /*signed_=*/false);
        }
        regs.write(instr.rd, value);

    } else if (m == "sb" || m == "sh" || m == "sw") {
        std::uint32_t address = to_unsigned(rs1 + instr.imm);
        if (m == "sb") {
            memory.write_byte(address, rs2);
        } else if (m == "sh") {
            memory.write_half(address, rs2);
        } else {
            memory.write_word(address, rs2);
        }

    } else if (m == "beq" || m == "bne" || m == "blt" || m == "bge" || m == "bltu" || m == "bgeu") {
        bool taken = (m == "beq" && rs1 == rs2) || (m == "bne" && rs1 != rs2) ||
                     (m == "blt" && rs1_s < rs2_s) || (m == "bge" && rs1_s >= rs2_s) ||
                     (m == "bltu" && rs1 < rs2) || (m == "bgeu" && rs1 >= rs2);
        if (taken) {
            pc_next = to_unsigned(static_cast<std::int64_t>(pc) + instr.imm);
        }

    } else if (m == "lui") {
        regs.write(instr.rd, instr.imm);
    } else if (m == "auipc") {
        regs.write(instr.rd, to_unsigned(static_cast<std::int64_t>(pc) + instr.imm));

    } else if (m == "jal") {
        regs.write(instr.rd, pc_next);
        pc_next = to_unsigned(static_cast<std::int64_t>(pc) + instr.imm);
    } else if (m == "jalr") {
        regs.write(instr.rd, pc_next);
        pc_next = to_unsigned(rs1 + instr.imm) & ~0x1u;

    } else if (m == "fence") {
        // no-op
    } else if (m == "ecall") {
        status = Status::Halted;
        halt_reason = "ecall";
        exit_code = regs.read_signed(10);
    } else if (m == "ebreak") {
        status = Status::Halted;
        halt_reason = "ebreak";
        exit_code = regs.read_signed(10);
    } else {
        throw IllegalInstructionError("unimplemented mnemonic '" + m + "'");
    }

    pc = pc_next;
}

}  // namespace riscv_sim
